src/syft/lib/python/primitive_factory.py

```diff
@@ -24,44 +26,74 @@ def isprimitive(value: Any) -> bool:
     return False
 
 
-class PrimitiveFactory(ABC):
+class PrimitiveFactory:
+    KNOWN_TYPES: Dict[
+        Union[
+            Type[int],
+            Type[float],
+            Type[bool],
+            Type[complex],
+            Type[list],
+            Type[str],
+            None,
+            NotImplementedType,
+        ],
+        Type[PyPrimitive],
+    ] = {}
+
+    DEFAULT_TYPE: Optional[Type[PyPrimitive]] = None
+
     @staticmethod
     @syft_decorator(typechecking=True)
     def generate_primitive(
-        value: Union[int, float, bool, complex, list, str, None, type(NotImplemented)],  # type: ignore
+        value: Union[int, float, bool, complex, list, str, None, NotImplementedType],
         id: Optional[UID] = None,
-    ) -> Union[PyPrimitive, type(NotImplemented)]:  # type: ignore
-        # syft relative
-        from .bool import Bool
-        from .complex import Complex
-        from .float import Float
-        from .int import Int
-        from .list import List
-        from .none import SyNone
-        from .string import String
-
-        if isinstance(value, bool):
-            return Bool(value=value, id=id)
+    ) -> Union[PyPrimitive, NotImplementedType]:
 
-        if isinstance(value, int):
-            return Int(value=value, id=id)
+        if value is NotImplemented:
+            return NotImplemented
 
-        if isinstance(value, float):
-            return Float(value=value, id=id)
+        key = value if value is None else type(value)
+        constructor = PrimitiveFactory.KNOWN_TYPES.get(
+            key, PrimitiveFactory.DEFAULT_TYPE
+        )
 
+        kwargs: Dict[str, Any]
         if isinstance(value, complex):
-            return Complex(real=value.real, imag=value.imag, id=id)
+            kwargs = {"real": value.real, "imag": value.imag, "id": id}
+        else:
+            kwargs = {"value": value, "id": id}
 
-        if type(value) is complex:
-            return String(value=value, id=id)
+        if constructor is None:
+            raise TypeError(
+                f"Type sent {key} not found in the registered types {PrimitiveFactory.KNOWN_TYPES.keys()} "
+                "and there is no default type constructor registered"
+            )
 
-        if type(value) is list:
-            return List(value=value, id=id)
+        return constructor(**kwargs)
 
-        if type(value) is str:
-            return String(value=value, id=id)
-        if value is NotImplemented:
-            return value
+    @staticmethod
+    @syft_decorator(typechecking=True)
+    def register_primitive(
+        python_primitive: Union[
+            Type[int],
+            Type[float],
+            Type[bool],
+            Type[complex],
+            Type[list],
+            Type[str],
+            None,
```

Comment:
Looks like we can use Type[None] to get typing.Type[NoneType]

Reply:
Actually, I remember trying to resolve this and just got into a loop of mypy errors that make no sense, so feel free to leave it as is.